8276553: ListView scrollTo() is broken after fix for JDK-8089589

johanvos opened the pull request:
After (re)setting the number of elements, make sure to do at least some estimation of the total size.
Added a testcase for this scenario.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276553](https://bugs.openjdk.java.net/browse/JDK-8276553): ListView scrollTo() is broken after fix for JDK-8089589


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Michael Strauß](https://openjdk.java.net/census#mstrauss) (@mstr2 - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/683/head:pull/683` \
`$ git checkout pull/683`

Update a local copy of the PR: \
`$ git checkout pull/683` \
`$ git pull https://git.openjdk.java.net/jfx pull/683/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 683`

View PR using the GUI difftool: \
`$ git pr show -t 683`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/683.diff">https://git.openjdk.java.net/jfx/pull/683.diff</a>

</details>

mstr2 commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:
```diff
@@ -898,6 +898,8 @@ public String getName() {
         resetSizeEstimates();
+        recalculateAndImproveEstimatedSize(2);
+
         cellCount.set(value);
```
Can this be implemented in a way that doesn't violate the property contract? Since this property is public API, `setCellCount(int)` should just call `cellCountProperty().set(int)`. Maybe you can extract the composite operation here into a private method and use that instead of `setCellCount(int)`.

`position` has the same problem.

kleopatra replied:
good catch! the "additional" stuff typically is done in the properties' invalidated ..

kevinrushforth replied:
Exactly. This ensures consistent behavior regardless of how a property value is set.

kevinrushforth replied:
I should add that this looks like a preexisting problem, but one that would be good to fix if possible.

johanvos replied:
Thanks for catching. I moved the additional code into the invalidated block, at the appropriate places.
Note that this caused some changes in the flow at runtime, where the number of evaluations of `updateItem` changed. Those changes are of constant order (and not linear with the amount of cells), and I made some changes to the test as they depended on hard numbers.

aghaisas commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:
```diff
@@ -1121,7 +1121,7 @@ public void updateItem(String color, boolean empty) {
                 Platform.runLater(() -> {
                     Toolkit.getToolkit().firePulse();
-                    assertEquals(0, rt_35395_counter);
+                    assertTrue(rt_35395_counter < 7);
```
I see that you have modified assertions to use "lesser than" some expected value. This may hide some incorrect test outcomes.
Along with "lesser than" assertion, do you think we should add a "greater than" assertion as well? This will have a range bounded expected value.
This is applicable for all assertion changes in this PR.

johanvos replied:
The hard values have been changed a number of times, and I believe it is not really a good metric. 
What we want to ensure is that there is no functional regression and no performance penalty. The tests calculate the number of updateItem invocations and require those to be a strict number. With JDK-8089589, we fixed a number of issues that are related to the fact that the total size of the view (in case all cells are rendered with their preferred height) is not known. This is done by using an estimate of the total size. The estimate is 100% correct if we call updateItem for every item, but that would lead to a huge performance penalty in case the list contains a large amount of items with equal height.
Hence, there is a balance between minimizing the updateItem calls but still have a fair estimate of the total dimensions. Rather than requiring that the amount of calls should be a fixed number, I think it makes more sense to ensure that the amount of calls stays within reasonable boundaries, and is not growing exponentially when we add linearly more items, for example.

aghaisas replied:
> The hard values have been changed a number of times, and I believe it is not really a good metric.

I agree completely.

> Rather than requiring that the amount of calls should be a fixed number, I think it makes more sense to ensure that the amount of calls stays within reasonable boundaries, and is not growing exponentially when we add linearly more items, for example.

My point is exactly this. I see that as part of this PR, you have added the upper boundary (rather than a fixed number) for assertions. I am asking whether we need a lower boundary as well?

kevinrushforth replied:
And in this case, since the value used to be 0 before this fix, I think that's fine. For the other cases would it make sense to have a `> 0` check? Or are those cases for which 0 would be an OK value?

aghaisas commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:
```diff
@@ -856,6 +856,8 @@ public String getName() {
         @Override protected void invalidated() {
             int cellCount = get();
+            resetSizeEstimates();
+            recalculateAndImproveEstimatedSize(2);
```
We can use recalculateEstimatedSize() instead of this method.
The effect is the same improvement with a size of 2.

johanvos replied:
True, good suggestion. I pushed that change.

kevinrushforth reviewed (approved):
The fix and new test looks fine. I added one comment/question regarding the existing tests that check a count, but I'm OK with either answer.

Approved.

mstr2 commented:
Would it be a good idea to fix the implementation of the `position` property as part of this PR (it has the same problem as `cellCount` before the fix)? If not, a follow-up ticket should be filed.

kevinrushforth commented:
/reviewers 2

kevinrushforth commented:
> Would it be a good idea to fix the implementation of the `position` property as part of this PR (it has the same problem as `cellCount` before the fix)? If not, a follow-up ticket should be filed.

A follow-up issue, [JDK-8278064](https://bugs.openjdk.java.net/browse/JDK-8278064), was filed a couple days ago.

johanvos commented:
> The fix and new test looks fine. I added one comment/question regarding the existing tests that check a count, but I'm OK with either answer.

The suggestion to test for positive values of the counters is a good idea. The lower this counter the better, but if the counter is 0, this means there has been no evaluation of updateItem. This should cause other tests to fail, but it helps if it would trigger a failure in this test as well.
I added these asserts on the places where previously the expected value was not 0.

johanvos commented:
/integrate
